addons/controllers/tanzuclusterbootstrap_controller.go

```diff
@@ -0,0 +1,588 @@
+		if secret, err := updateValuesFromSecret(bootstrap.Spec.CPI.ValuesFrom, packageShortName(bootstrap.Spec.CPI.RefName)); err != nil {
+			return nil, nil, err
+		} else { //nolint:revive
+			createdSecrets = append(createdSecrets, secret)
+		}
+		if provider, err := updateValuesFromProvider(bootstrap.Spec.CPI.ValuesFrom, packageShortName(bootstrap.Spec.CPI.RefName)); err != nil {
+			return nil, nil, err
+		} else { //nolint:revive
+			createdProviders = append(createdProviders, provider)
+		}
+	}
+
+	if bootstrap.Spec.CSI != nil && bootstrap.Spec.CSI.ValuesFrom != nil {
+		if secret, err := updateValuesFromSecret(bootstrap.Spec.CSI.ValuesFrom, packageShortName(bootstrap.Spec.CSI.RefName)); err != nil {
+			return nil, nil, err
+		} else { //nolint:revive
+			createdSecrets = append(createdSecrets, secret)
+		}
+		if provider, err := updateValuesFromProvider(bootstrap.Spec.CSI.ValuesFrom, packageShortName(bootstrap.Spec.CSI.RefName)); err != nil {
+			return nil, nil, err
+		} else { //nolint:revive
+			createdProviders = append(createdProviders, provider)
+		}
+	}
+
+	if bootstrap.Spec.Kapp != nil && bootstrap.Spec.Kapp.ValuesFrom != nil {
+		if secret, err := updateValuesFromSecret(bootstrap.Spec.Kapp.ValuesFrom, packageShortName(bootstrap.Spec.Kapp.RefName)); err != nil {
+			return nil, nil, err
+		} else { //nolint:revive
+			createdSecrets = append(createdSecrets, secret)
+		}
+		if provider, err := updateValuesFromProvider(bootstrap.Spec.Kapp.ValuesFrom, packageShortName(bootstrap.Spec.Kapp.RefName)); err != nil {
+			return nil, nil, err
+		} else { //nolint:revive
+			createdProviders = append(createdProviders, provider)
+		}
+	}
+
+	for _, additionalPackage := range bootstrap.Spec.AdditionalPackages {
+		if additionalPackage.ValuesFrom != nil {
+			if additionalPackage.ValuesFrom.SecretRef != "" {
+				if secret, err := updateValuesFromSecret(additionalPackage.ValuesFrom, packageShortName(additionalPackage.RefName)); err != nil {
+					return nil, nil, err
+				} else { //nolint:revive
+					createdSecrets = append(createdSecrets, secret)
+				}
+			}
+			if additionalPackage.ValuesFrom.ProviderRef != nil {
+				if provider, err := updateValuesFromProvider(additionalPackage.ValuesFrom, packageShortName(additionalPackage.RefName)); err != nil {
+					return nil, nil, err
+				} else { //nolint:revive
+					createdProviders = append(createdProviders, provider)
+				}
+			}
+		}
+	}
+
+	return createdSecrets, createdProviders, nil
+}
+
+// ensureOwnerRef will ensure the provided OwnerReference onto the secrets and provider objects
+func (r *tanzuClusterBootstrapReconciler) ensureOwnerRef(ownerRef *metav1.OwnerReference, secrets []*corev1.Secret, providers []*unstructured.Unstructured) error {
+	for _, secret := range secrets {
+		secret.OwnerReferences = clusterapiutil.EnsureOwnerRef(secret.OwnerReferences, *ownerRef)
+		// TODO patch
+		if err := r.Client.Update(r.context, secret); err != nil {
```

Comment:
suggestion: Use PatchSet (defined in pkg/v1/util/patchset/patchset.go). For usage example, see pkg/v1/tkr/controllers/source/tkr_source_controller.go